### DISCUSSION TITLE
No labels in main menu

### DISCUSSION THREAD

**Discussion by Vasmares** [BimmerGestalt/AAIdrive#430] [Q&A] [answered]
Hi,

i came to notice that for some entries in the connected drive menu, the is no label, just the icons.
Also the addons dont have a label.

Seen also here in the readme:
![labels](https://user-images.githubusercontent.com/10863304/136272780-8d4ef6d9-b48d-415b-97c5-c8ba6e857a7d.png)

Is this intended or are there labels / translations missing ?

**Comment by hufman** [comment, marked as answer]
Indeed, this is intentional! Those main menu entries use icons and labels from resource packs that are cryptographically locked to the app cert that is used to log in to the car. These main menu entries are additionally restricted in that they are unable to change to point to a different resource from the resource packs, they are hardcoded. This app in the screenshot, which is used for the Notifications functionality, was selected for its widget functionality (input and popups in ID4) and it happened to have a relevant icon. The text pack for this app has two entries that would not be helpful as an app label (`<APPTITLE>` and `Error: app doesn't work.`), but actually the layout is hardcoded to point to a missing resource. This resource is meant to be used for official app integrations (Samsung Smartthings is the only known integration of this type) which would have been expected to provide their own BMW-signed text packs with the proper string in place.

Even for apps that do have proper text packs set up (the Smartthings app is used for the Maps and Screen Mirroring functionality, for example), AAIdrive just doesn't sent the text pack to the car, to purposely keep the app label blank instead of being mistaken for the official app.

The other icons you find (the individual music apps in the Media list, the Notifications app in the Office section) are of a different type which only act as placeholder icons. These can have entirely custom icons and labels, not bound by a signed resource pack, but can not hold any UI within and can only act as buttons. Originally they are intended to be clicked, show a loading spinner while the full app is initialized, and then continue into the full app. AAIdrive uses them as simple button triggers (to switch the music source, for example) and then ask the car to open the primary app, which unfortunately seems unreliable in ID4.

Let me know if I can explain anything further!

**Reply by Vasmares** [reply]
Thaks for pointing that out.
In terms "layout", would it be possible to put all this Icons / App icons into some kind of "Sub Menu" ?

Like adding one menu Item, and placing all the options in a sub menu ?
Maybe this would also fix the "icon but no label" menu item problem ?

![148870-cars-feature-mini-system-2-image2-yzhzqx5c7a](https://user-images.githubusercontent.com/10863304/137217309-a434a5f3-ed57-475b-bf31-52b8890a7d0d.jpg)

**Reply by hufman** [reply]
As far as I know, the car API does not provide such a sub-menu functionality. The `ui_description.xml` specifies which overall menu to add the app's entry button, using the [same categories](https://github.com/BimmerGestalt/AAIdrive/blob/9acbe8aa6c44c3c7fc9c3b3166a0de9db4e5c2a1/app/src/main/java/me/hufman/androidautoidrive/carapp/AMAppList.kt#L10) as the AM Icon API, and this `ui_description.xml` resource is signed by the BMW and can't be modified to use a different location.
